xcode/Subconscious/Shared/Components/Common/Profile/EditProfileSheet.swift

```diff
@@ -0,0 +1,268 @@
+//
+//  EditProfileSheet.swift
+//  Subconscious
+//
+//  Created by Ben Follington on 24/4/2023.
+//
+
+import os
+import ObservableStore
+import SwiftUI
+import Combine
+
+enum EditProfileSheetAction: Equatable {
+    case populate(UserProfileEntry?)
+    case nicknameField(FormFieldAction<String>)
+    case bioField(FormFieldAction<String>)
+    case pfpUrlField(FormFieldAction<String>)
+}
+
+private struct NicknameFieldCursor: CursorProtocol {
+    typealias Model = EditProfileSheetModel
+    typealias ViewModel = FormField<String, Petname>
+
+    static func get(state: Model) -> ViewModel {
+        state.nicknameField
+    }
+    
+    static func set(state: Model, inner: ViewModel) -> Model {
+        var model = state
+        model.nicknameField = inner
+        return model
+    }
+    
+    static func tag(_ action: ViewModel.Action) -> Model.Action {
+        .nicknameField(action)
+    }
+}
+
+private struct BioFieldCursor: CursorProtocol {
+    typealias Model = EditProfileSheetModel
+    typealias ViewModel = FormField<String, String>
+
+    static func get(state: Model) -> ViewModel {
+        state.bioField
+    }
+    
+    static func set(state: Model, inner: ViewModel) -> Model {
+        var model = state
+        model.bioField = inner
+        return model
+    }
+    
+    static func tag(_ action: ViewModel.Action) -> Model.Action {
+        .bioField(action)
+    }
+}
+
+private struct PfpUrlFieldCursor: CursorProtocol {
+    typealias Model = EditProfileSheetModel
+    typealias ViewModel = FormField<String, URL>
+
+    static func get(state: Model) -> ViewModel {
+        state.pfpUrlField
+    }
+    
+    static func set(state: Model, inner: ViewModel) -> Model {
+        var model = state
+        model.pfpUrlField = inner
+        return model
+    }
+    
+    static func tag(_ action: ViewModel.Action) -> Model.Action {
+        .pfpUrlField(action)
+    }
+}
+
+struct EditProfileSheetEnvironment {
```

Comment:
If we don't plan to do anything with this soon, it could be `Void` instead. When calling update, you pass `()` for void.

Reply:
👍 changed for `FormFieldEnvironment` as well. 

Fun fact, you can do:
```swift
typealias FormFieldEnvironment = Void

let a: FormFieldEnvironment = FormFieldEnvironment()
let b: FormFieldEnvironment = Void()
let c: FormFieldEnvironment = ()
```

And they all work!